torch/fx/utils.py

```diff
@@ -0,0 +1,144 @@
+
+        for arg in node.args:
+            if isinstance(arg, torch.fx.Node):
+                use_set.add(arg)
+        for k, v in node.kwargs.items():
+            if isinstance(v, torch.fx.Node):
+                use_set.add(v)
+        def_set.add(node)
+
+    block_inputs = block_uses - block_defs
+    block_outputs = list(rest_uses - rest_defs)
+    assert len(block_outputs) == 1
+
+    # ===== Stage 2: Create submodule for extracted module =====
+    # TODO: ideally we would just symbolically trace `target_module` but
+    # I have to think through how to associate the positional/kwarg
+    # inputs with traced values
+
+    submod_graph = torch.fx.Graph()
+    submod_delegate = torch.fx.DefaultDelegate(target_module, submod_graph)
+
+    # Map Node in the original Graph to the node in the new graph
+    remap_table = {}
+    for block_input in block_inputs:
+        remap_table[block_input] = submod_delegate.placeholder(block_input.name)
+
+    for node in block_nodes:
+        def arg_transform(n : torch.fx.Node):
+            return remap_table[n]
+
+        # We need to fixup all qualfied names to strip off the base
+        # qualname, since now all qualnames in the new Module should
+        # be wrt that new module
+        def qualname_transform(qualname : str) -> str:
+            return qualname.replace(f'{target_qualname}.', '')
+
+        submod_graph_node = submod_graph.node_copy(node, arg_transform, qualname_transform)
+
+        # Keep track of created output values
+        remap_table[node] = submod_graph_node
+
+    # Register output
+    submod_graph.output(remap_table[block_outputs[0]])
+    submod_module = torch.fx.GraphModule(target_module, submod_graph)
+
+    # ===== Stage 3: Create new base GraphModule with submodule nodes replaced
+    #       with a callsite to that new submodule =====
+
+    base_graph = torch.fx.Graph()
+    base_delegate = torch.fx.DefaultDelegate(mod, base_graph)
+    base_remap_table = {}
+
+    callsite_inserted = False
+    for orig_node in mod.graph.nodes:
+        def arg_transform(n : torch.fx.Node):
+            return base_remap_table[n]
+        if orig_node in block_defs:
+            if not callsite_inserted:
+                args = tuple(arg_transform(n) for n in block_inputs)
+                base_remap_table[block_outputs[0]] = base_delegate.create_node(
+                    'call_module', target_qualname, args, {}, block_outputs[0].name,
+                    orig_node.module_qualname)
+                callsite_inserted = True
+            # Fallthrough. For each node that's a def in the subgraph,
+            # just elide copying it.
+        else:
+            base_remap_table[orig_node] = base_graph.node_copy(orig_node, arg_transform)
+
+    # Register output
+    base_graph.output(base_remap_table[mod.graph.result])
+
+    # NOT GREAT: deepcopy root so we can mutate it and insert our new
+    # submodule
+
+    new_root = copy.deepcopy(mod)
```

Comment:
so `GraphModule` these days copies referenced sub-components of the root, right? can't we just create GraphModule first and then re-assign the `new_root_target` afterwards? I think it'd be safe and avoids the copy.

Reply:
Yeah, that's probably a good idea. This was a dumb merge conflict resolution after rebasing past zach's PR